pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -98,9 +98,10 @@ func (pl *DefaultPreemption) PostFilter(ctx context.Context, state *framework.Cy
 // using the nominated resources and the nominated pod could take a long time
 // before it is retried after many other pending pods.
 func preempt(ctx context.Context, fh framework.FrameworkHandle, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) (string, error) {
-	cs := fh.ClientSet()
-	// TODO(Huang-Wei): get pod from informer cache instead of API server.
-	pod, err := util.GetUpdatedPod(cs, pod)
+	// It's safe to directly fetch pod here. Because the informer cache has already been
+	// initialized when creating the Scheduler obj, i.e., factory.go#MakeDefaultErrorFunc().
+	// However, tests may need to manually initialize the shared pod informer.
+	pod, err := fh.SharedInformerFactory().Core().V1().Pods().Lister().Pods(pod.Namespace).Get(pod.Name)
```

Comment:
To avoid initialization issues in the future, let's stick with the practice of getting the lister in New.

Reply:
It's now already initialized in New, indirectly  - the full call chain is like this: scheduler.go#New() -> factory.go#create() -> factory.go#MakeDefaultErrorFunc().

Reply:
I guess what I am suggesting is to establish a pattern in all plugins to not access listers directly like this, but rather create a variable in the plugin that we set in New.

Reply:
I suppose you're suggesting:

```go
informerFactory....Lister()
```

in New(), right? Just like what we did for PDBLister.

Reply:
Here is my suggestion:

```
type DefaultPreemption struct {
	fh framework.FrameworkHandle
        podLister  corelisters.PodLister 
        pdbLister v1beta1.PodDisruptionBudgetLister
}
```

This line changes to:
```
pod, err := pl.podLister.Pods(pod.Namespace).Get(pod.Name)
```

and New:
```
// New initializes a new plugin and returns it.
func New(_ runtime.Object, fh framework.FrameworkHandle) (framework.Plugin, error) {
	var pdbLister v1beta1.PodDisruptionBudgetLister
	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.PodDisruptionBudget) {
		// A hack to initialize pdbLister in sharedInformerFactory.
		pdbLister = fh.SharedInformerFactory().Policy().V1beta1().PodDisruptionBudgets().Lister()
	}

         pl := DefaultPreemption{
            fh: fh, 
            podLister: SharedInformerFactory().Core().V1().Pods().Lister().Pods(pod.Namespace).Get(pod.Name),
            pdbLister: pdbLister,
         }
	return &pl, nil
}
```
and lastly

```
  func getPodDisruptionBudgets(pdbLister v1beta1.PodDisruptionBudgetLister) ([]*policy.PodDisruptionBudget, error) {
	if pdbLister != nil {
		return pdbLister.List(labels.Everything())
	}
	return nil, nil
}
```

Reply:
I think we discussed it before. My option is that the xyzLister can be fetched from the informerFactory, so we just need to ensure it's initialized properly, and then just keep the informerFactory variable there.

We can move the discussion to #92781.

Reply:
Should the method 'SharedInformerFactory()' should be removed from 'framework.FrameworkHandle'? Feel like the people who want to write their own plugin are easily making same mistake.

Reply:
> Should the method 'SharedInformerFactory()' should be removed from 'framework.FrameworkHandle'?

Big no... It's the canonical way to interface with API resources.

> Feel like the people who want to write their own plugin are easily making same mistake.

We should document it in our development doc. Ideally in https://github.com/kubernetes-sigs/scheduler-plugins/

Reply:
> > Should the method 'SharedInformerFactory()' should be removed from 'framework.FrameworkHandle'?
> 
> Big no... It's the canonical way to interface with API resources.

Sorry, I'm not clear my thought. Actually I'm thinking of removing the SharedInformerFactory from the framework.Framworkhandle, but add a new parameter to the New() method. As my understanding, we shouldn't allow extension point to use informer factory, Any lister should be getting at the New() method. But yes, even if we just pass the informer factory to the New() method, people still can pass that through the data structure of plugin itself, just like frameworkhandle. In the end, the user still could be using it with wrong way. So probably there needs another more complex solution here. Forgive me if I'm still thinking something wrong, still trying to learn here :)

> 
> > Feel like the people who want to write their own plugin are easily making same mistake.
> 
> We should document it in our development doc. Ideally in https://github.com/kubernetes-sigs/scheduler-plugins/

Yea, that works. just thinking of there should be something in the design to prevent people make mistakes.